locale.c

```diff
@@ -147,6 +147,81 @@ static int debug_initialization = 0;
 #define PERL_IN_LOCALE_C
 #include "perl.h"
 
+#ifdef USE_FAKE_OLD_MINGW_LOCALES
+
+   /* Use -Accflags=-DUSE_FAKE_OLD_MINGW_LOCALES on a POSIX or *nix box to get
+    * a semblance of pretending the locale handling is that of MingW which
+    * doesn't use UCRT (hence 'OLD' in the name).  This allows for finding many
+    * (but certainly not all) bugs on such a system without accessing a real
+    * MingW box. */
+
+#  define WIN32
+#  undef P_CS_PRECEDES
+#  undef CURRENCY_SYMBOL
+#  define CP_UTF8 -1
+#  undef _configthreadlocale
+#  define _configthreadlocale(arg) NOOP
+
+#  define MultiByteToWideChar(cp, flags, byte_string, m1, wstring, req_size) \
+                    (mbsrtowcs(wstring, &byte_string, req_size, NULL) + 1)
+#  define WideCharToMultiByte(cp, flags, wstring, m1, byte_string,          \
+                              req_size, default_char, found_default_char)   \
+                    (wcsrtombs(byte_string, &wstring, req_size, NULL) + 1)
+
+#  ifdef USE_LOCALE
+
+static const wchar_t * wsetlocale_buf = NULL;
+static Size_t wsetlocale_buf_size = 0;
+
+STATIC
+const wchar_t *
+S_wsetlocale(const int category, const wchar_t * wlocale)
+{
+    /* Windows uses a setlocale that takes a wchar_t* locale.  Other boxes
+     * don't have this, so this Windows replacement converts the wchar_t input
+     * to plain 'char*', calls plain setlocale(), and converts the result back
+     * to 'wchar_t*' */
+
+    const char * byte_locale = NULL;
+    if (wlocale) {
+        byte_locale = Win_wstring_to_byte_string(CP_UTF8, wlocale);
+    }
+
+    const char * byte_result = setlocale(category, byte_locale);
+    Safefree(byte_locale);
+    if (byte_result == NULL) {
+        return NULL;
+    }
+
+    const wchar_t * wresult = Win_byte_string_to_wstring(CP_UTF8, byte_result);
+
+    if (! wresult) {
+        return NULL;
+    }
+
+    /* Emulate a global static memory return from wsetlocale().  This currently
+     * leaks at process end; would require changing LOCALE_TERM to fix that */
+    Size_t string_size = wcslen(wresult) + sizeof(wchar_t);
```

Comment:
I don't think so.  But maybe I'm not understanding what's going on.  The extra space is for a wide NULL.  That is bigger than 1.  Since the data structure is a wchar_t, the size allocated must be a multiple of sizeof(wchar_t).

Reply:
wcslen() returns the number of wide characters in the string, which this code then adds a number of bytes to, if this is space for a wide `NUL`, then it should be one, since you're only adding a single wide character.

Newx() and Copy() below then correctly turn that wide character count into a byte count.

It could be you intend string_size to be a byte count, but that doesn't seem likely given Newx() and Copy() use them as wide character counts.

Reply:
Now I understand